libsupport/include/katana/ProgressTracer.h

```diff
@@ -0,0 +1,222 @@
+  uint32_t GetHostID() { return host_id_; }
+  uint32_t GetNumHosts() { return num_hosts_; }
+
+  /// Close is called when a tracer is finished processing spans
+  /// This should be called to ensure any and all buffered spans are
+  /// flushed
+  virtual void Close() = 0;
+
+private:
+  static std::unique_ptr<ProgressTracer> tracer_;
+
+  std::shared_ptr<ProgressSpan> active_span_ = nullptr;
+  uint32_t host_id_;
+  uint32_t num_hosts_;
+  std::shared_ptr<ProgressSpan> default_active_span_ = nullptr;
+};
+
+class KATANA_EXPORT ProgressScope {
+  friend class ProgressTracer;
+
+  ProgressScope(std::shared_ptr<ProgressSpan> span) : span_(span) {}
+
+public:
+  ~ProgressScope();
+  ProgressScope(const ProgressScope&) = delete;
+  ProgressScope(ProgressScope&&) = delete;
+  ProgressScope& operator=(const ProgressScope&) = delete;
+  ProgressScope& operator=(ProgressScope&&) = delete;
+
+  /// Get the scope’s underlying span to add tagging/logging,
+  /// span relationships, and handle multiple active spans at a time
+  ProgressSpan& span() { return *span_; }
+
+  /// Closes the underlying ProgressSpan if the ProgressScope was created
+  /// with the flag finish_on_close=true
+  /// Note that this will only close the underlying ProgressSpan when all
+  /// of its active children ProgressSpans have been finished
+  /// This will be called by RAII if it has not already been called
+  void Close();
+
+private:
+  std::shared_ptr<ProgressSpan> span_;
+};
+
+class KATANA_EXPORT ProgressContext {
+public:
+  virtual ~ProgressContext() = default;
+  virtual std::unique_ptr<ProgressContext> Clone() const noexcept = 0;
+  virtual std::string GetTraceID() const noexcept;
+  virtual std::string GetSpanID() const noexcept;
+};
+
+class KATANA_EXPORT ProgressSpan {
+  virtual void Close() = 0;
+
+protected:
+  ProgressSpan(std::shared_ptr<ProgressSpan> parent) : parent_(parent) {}
+
+public:
+  /// If Finish has not already been called for the ProgressSpan, it's
+  /// destructor must do so.
+  virtual ~ProgressSpan() = default;
+
+  /// Adds a tag to the span.
+  virtual void SetTags(const Tags& tags) = 0;
+  void SetError() { SetTags({{"error", true}}); }
+
+  /// Output logging as well as standard metrics and extra stats
+  /// Current standard metrics: max_mem, mem, host, and timestamp.
+  virtual void Log(const std::string& message, const Tags& tags) = 0;
+  void Log(const std::string& message) { Log(message, {}); }
+  void LogError(const std::string& message) {
+    Log(message, {{"event", "error"}});
+  }
+  void LogError(const std::string& message, const ErrorInfo& error);
+
+  /// Get span's context for propagating across process boundaries
+  virtual const ProgressContext& GetContext() const noexcept = 0;
+
+  /// Primarily for internal class use only
+  void MarkScopeClosed();
+  virtual bool ScopeClosed();
+  bool IsFinished() { return finished_; }
+  const std::shared_ptr<ProgressSpan>& GetParentSpan() { return parent_; }
```

Comment:
`ProgressSpan* GetParentSpan() { return parent_.get(); }` ?

Reply:
We need the shared pointer for logic in the tracer class